platform/cxxsupport/TOOLCHAIN_ARMC5/tuple

```diff
@@ -0,0 +1,777 @@
+ * tuple can be constexpr either, so this is a C++11 implementation, not C++14.
+ *
+ * Allocators are not supported.
+ */
+#ifndef __tuple
+#define __tuple
+
+#include <mstd_utility>
+#include <type_traits>
+
+namespace std
+{
+
+template <typename... T>
+struct tuple;
+
+template <size_t I, class T>
+struct tuple_element;
+
+// ARM C 5 (incorrectly? doesn't allow duplicate using - utility has already done this (for pair)
+//template <size_t I, class T>
+//using tuple_element_t = typename tuple_element<I, T>::type;
+
+template <size_t I, typename Type0, typename... TypeN>
+struct tuple_element<I, tuple<Type0, TypeN...>> : tuple_element<I-1, tuple<TypeN...>> { };
+
+template <typename Type0, typename... TypeN>
+struct tuple_element<0, tuple<Type0, TypeN...>> : type_identity<Type0> { };
+
+template <size_t I, typename T>
+struct tuple_element<I, const T> : type_identity<add_const_t<tuple_element_t<I,T>>> { };
+
+template <size_t I, class T>
+struct tuple_element<I, volatile T> : type_identity<add_volatile_t<tuple_element_t<I,T>>> { };
+
+template <size_t I, class T>
+struct tuple_element<I, const volatile T> : type_identity<add_cv_t<tuple_element_t<I,T>>> { };
+
+template <size_t I, typename... T>
+tuple_element_t<I, tuple<T...>> &get(tuple<T...> &t) noexcept;
+
+template <size_t I, typename... T>
+tuple_element_t<I, tuple<T...>> &&get(tuple<T...> &&t) noexcept;
+
+template <size_t I, typename... T>
+const tuple_element_t<I, tuple<T...>> &get(const tuple<T...> &t) noexcept;
+
+template <size_t I, typename... T>
+const tuple_element_t<I, tuple<T...>> &&get(const tuple<T...> &&t) noexcept;
+
+namespace impl {
+
+/* Tuple element container - tuple has these as multiple base classes */
+template <size_t, typename T>
+struct tuple_elem {
+    constexpr tuple_elem() : value() { } // tuple default constructor value-initializes elements
+    explicit tuple_elem(const T &val) : value(val) { }
+    template <typename U, typename = enable_if_t<is_constructible<T,U&&>::value>>
+    explicit tuple_elem(U&& val) : value(std::forward<U>(val)) { }
+    tuple_elem(const tuple_elem &other) = default;
+    //tuple_elem(tuple_elem &&other) = default; // XXX ARMC5 does not support
+    tuple_elem(tuple_elem &&other) : value(std::move(other).value) { }
+    //tuple_elem &operator=(tuple_elem &&other) = default;// XXX ARMC5 does not support
+    tuple_elem &operator=(tuple_elem &&other) { value = std::move(other).value; }
+    tuple_elem &operator=(const T &val) { value = val; return *this; }
+    template <typename U, typename = enable_if_t<is_assignable<T &,U &&>::value>>
+    tuple_elem &operator=(U&& val) { value = std::forward<U>(val); return *this; }
+    void swap(T &other) { using namespace std; std::swap(value, other); }
+    T value;
+};
+
+template <typename Seq, typename... T>
+class tuple_base;
+
+template <typename T>
+struct is_tuple : std::false_type { };
+
+template <typename... T>
+struct is_tuple<tuple<int, T...>> : std::true_type { };
```

Comment:
I'm not sure to understand why we need the `int` or `index_sequence` as the first template parameter; a tuple can begin with a different template argument and even with no argument at all : `tuple<>`

Reply:
I think that was intended to be `tuple_base<index_sequence<I...>, T...>`